packages/blockchain-wallet-v4-frontend/src/data/components/interest/sagas.utils.ts

```diff
@@ -306,10 +307,108 @@ export default ({ coreSagas, networks }: { coreSagas: any; networks: any }) => {
     }
   }
 
+  const toCustodialDropdown = currencyDetails => {
+    // this object has to be equal to object we do expect in dropdown
+    const { coin, ...restDetails } = currencyDetails
+    return [
+      {
+        ...restDetails,
+        address: currencyDetails.address,
+        type: ADDRESS_TYPES.CUSTODIAL,
+        label: `${coin} Trading Wallet`
+      }
+    ]
+  }
+
+  const getCustodialAccountForCoin = function * (coin: CoinType) {
+    let defaultAccountR
+
+    yield put(actions.components.send.fetchPaymentsTradingAccount(coin))
+    yield take([
+      actionTypes.components.send.FETCH_PAYMENTS_TRADING_ACCOUNTS_SUCCESS,
+      actionTypes.components.send.FETCH_PAYMENTS_TRADING_ACCOUNTS_FAILURE
+    ])
+
+    const accountAddress = selectors.components.send.getPaymentsTradingAccountAddress(
+      coin,
+      yield select()
+    )
+    const state = yield select()
+
+    switch (coin) {
+      case 'BCH':
+        const bchAccount = selectors.components.simpleBuy
+          .getSBBalances(state)
+          .map(x => ({
+            ...x.BCH,
+            coin,
+            address: accountAddress ? accountAddress.data : null
+          }))
+          .map(toCustodialDropdown)
+        const bchDefaultIndex = yield select(
+          selectors.core.wallet.getDefaultAccountIndex
+        )
+        defaultAccountR = bchAccount.map(nth(bchDefaultIndex))
+        break
+      case 'BTC':
+        const btcAccounts = selectors.components.simpleBuy
+          .getSBBalances(state)
+          .map(x => ({
+            ...x.BTC,
+            address: accountAddress ? accountAddress.data : null
+          }))
+          .map(toCustodialDropdown)
+        const btcDefaultIndex = yield select(
+          selectors.core.wallet.getDefaultAccountIndex
+        )
+        defaultAccountR = btcAccounts.map(nth(btcDefaultIndex))
+        break
+      case 'ETH':
+        const ethAccount = selectors.components.simpleBuy
+          .getSBBalances(state)
+          .map(x => ({
+            ...x.ETH,
+            coin,
+            address: accountAddress ? accountAddress.data : null
+          }))
+          .map(toCustodialDropdown)
+        defaultAccountR = ethAccount.map(head)
+        break
+      case 'PAX':
+      case 'USDT':
+        const erc20Account = selectors.components.simpleBuy
+          .getSBBalances(state)
+          .map(x => ({
+            ...x[coin],
+            coin,
+            address: accountAddress ? accountAddress.data : null
+          }))
+          .map(toCustodialDropdown)
+        defaultAccountR = erc20Account.map(head)
+        break
+      case 'XLM':
+        const xlmAcccount = selectors.components.simpleBuy
+          .getSBBalances(state)
+          .map(x => ({
+            ...x.XLM,
+            coin,
+            address: accountAddress ? accountAddress.data : null
+          }))
+          .map(toCustodialDropdown)
+        defaultAccountR = xlmAcccount.map(head)
+        break
```

Comment:
all of logic in the switch cases are identical except for the coin map.  i think you make it more generic.  The only real check you need to do is if the coin is BTC or BCH, then you need to also select the `defaultAccountIndex`

Reply:
Yes will improve it ;)